File: gbcs-api/src/main/java/net/woggioni/gbcs/api/event/ResponseEvent.java
package net.woggioni.gbcs.api.event;

import io.netty.buffer.ByteBuf;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

sealed public abstract class ResponseEvent {
    @Getter
    @RequiredArgsConstructor
    public final static class ChunkReceived extends ResponseEvent {
        private final ByteBuf chunk;
    }

    public final static class NoContent extends ResponseEvent {
    }

    @Getter
    @RequiredArgsConstructor
    public final static class LastChunkReceived extends ResponseEvent {
        private final ByteBuf chunk;
    }

    @Getter
    @RequiredArgsConstructor
    public final static class ExceptionCaught extends ResponseEvent {
        private final Throwable cause;
    }
}

File: gbcs-server-memcached/src/main/kotlin/net/woggioni/gbcs/server/memcached/client/ResponseEvent.kt
package net.woggioni.gbcs.server.memcached.client

import java.nio.ByteBuffer

sealed interface ResponseEvent {
    class ResponseReceived(val response : MemcacheResponse) : ResponseEvent
    class ResponseContentChunkReceived(val chunk: ByteBuffer) : ResponseEvent
    class LastResponseContentChunkReceived(val chunk: ByteBuffer) : ResponseEvent
    class ExceptionCaught(val cause : Throwable) : ResponseEvent
}